src/mscorlib/src/System/String.CoreCLR.cs

```diff
@@ -0,0 +1,153 @@
+        //from native.
+        public static readonly String Empty;
+
+        // Gets the character at a specified position.
+        //
+        [System.Runtime.CompilerServices.IndexerName("Chars")]
+        public extern char this[int index]
+        {
+            [MethodImpl(MethodImplOptions.InternalCall)]
+            get;
+        }
+
+        // Gets the length of this string
+        //
+        /// This is a EE implemented function so that the JIT can recognise is specially
+        /// and eliminate checks on character fetchs in a loop like:
+        ///        for(int i = 0; i < str.Length; i++) str[i]
+        /// The actually code generated for this will be one instruction and will be inlined.
+        //
+        public extern int Length
+        {
+            [MethodImplAttribute(MethodImplOptions.InternalCall)]
+            get;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern static String FastAllocateString(int length);
+
+        // Is this a string that can be compared quickly (that is it has only characters > 0x80 
+        // and not a - or '
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool IsFastSort();
+        // Is this a string that only contains characters < 0x80.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool IsAscii();
+
+#if FEATURE_COMINTEROP
+        // Set extra byte for odd-sized strings that came from interop as BSTR.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern void SetTrailByte(byte data);
+        // Try to retrieve the extra byte - returns false if not present.
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern bool TryGetTrailByte(out byte data);
+#endif
+
+        public static String Intern(String str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            return Thread.GetDomain().GetOrInternString(str);
+        }
+
+        public static String IsInterned(String str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            return Thread.GetDomain().IsStringInterned(str);
+        }
+
+        // Copies the source String (byte buffer) to the destination IntPtr memory allocated with len bytes.
+        internal unsafe static void InternalCopy(String src, IntPtr dest, int len)
+        {
+            if (len == 0)
+                return;
+            fixed (char* charPtr = &src._firstChar)
+            {
+                byte* srcPtr = (byte*)charPtr;
+                byte* dstPtr = (byte*)dest;
+                Buffer.Memcpy(dstPtr, srcPtr, len);
+            }
+        }
+
+        unsafe internal int GetBytesFromEncoding(byte* pbNativeBuffer, int cbNativeBuffer, Encoding encoding)
```

Comment:
Nit, these should be "internal unsafe" per coding guidelines. again, was that way before.